lib/services/consumed/autocomplete.js

```diff
@@ -0,0 +1,130 @@
+/* @flow */
+import { CompositeDisposable, Disposable } from "atom";
+
+import type { Store } from "../../store";
+import type WatchesStore from "../../store/watches";
+import type WatchStore from "../../store/watch";
+
+/**
+ * This acts as a global storage for the consumed service.
+ */
+export class AutocompleteWatchEditor {
+  disposables = new CompositeDisposable();
+  /** The `consumed autocompleteWatchEditor` */
+  addAutocompleteToEditor = (
+    editor: atom$TextEditor,
+    labels: Array<string>
+  ) => {
+    return;
+  };
+
+  /**
+   * This function is called on activation of autocomplete, or if autocomplete is
+   * already active, then it is called when hydrogen activates.
+   *
+   * @param {Store} store - The global Hydrogen store.
+   * @param {Function} watchEditor - The function provided by `autocomplete.watchEditor`.
+   * @returns {Disposable} This is for clean up when autocomplete or hydrogen deactivate.
+   */
+  consume(store: Store, watchEditor: atom$AutocompleteWatchEditor) {
+    this.addAutocompleteToEditor = watchEditor;
+
+    // Add autocomplete capabilities to already existing watches
+    for (let kernel of store.runningKernels) {
+      for (let watch of kernel.watchesStore.watches) {
+        this.addAutocompleteToWatch(kernel.watchesStore, watch);
+      }
+      this.disposables.add(kernel.watchesStore.autocompleteDisposables);
+    }
+
+    const disposable = new Disposable(() => this.disable(store));
+    store.subscriptions.add(disposable);
+    return disposable;
+  }
+
+  /**
+   * This function is just for cleaning up when either autocomplete or hydrogen is deactivating.
+   *
+   * @param {Store} store - The global Hydrogen store.
+   */
+  disable(store: Store) {
+    // Removes the consumed function `watchEditor`
+    this.addAutocompleteToEditor = (
```

Comment:
I think something like
```js
if (this.addAutocompleteToEditor) this.addAutocompleteToEditor.dispose();
```
would do the trick ?

Reply:
If you look at the type declaration of `addAutocompleteToEditor` at the top of the class its a function. I am replacing the function I consumed with one that does nothing to avoid doing null/undefined checks everywhere it is used.

Reply:
Yeah, but if you `this.addAutocompleteToEditor.dispose();`, I believe the actual function object would stay there and behaves like the function returning `null` on call.

Reply:
Dispose doesn't exist in the function prototype

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype

Is it apart of something else?

Reply:
As far as I know dispose is a custom method of atom disposable and compositedisposable class

Reply:
NVM. I misunderstood that `this.addAutoCompleteToEditor` itself is disposable.